# Advent of Code 2020 - Day 2

### Challenge

-------------

--- Day 2: Password Philosophy ---
Your flight departs in a few days from the coastal airport; the easiest way down to the coast from here is via toboggan.

The shopkeeper at the North Pole Toboggan Rental Shop is having a bad day. "Something's wrong with our computers; we can't log in!" You ask if you can take a look.

Their password database seems to be a little corrupted: some of the passwords wouldn't have been allowed by the Official Toboggan Corporate Policy that was in effect when they were chosen.

To try to debug the problem, they have created a list (your puzzle input) of passwords (according to the corrupted database) and the corporate policy when that password was set.

For example, suppose you have the following list:

`1-3 a: abcde
1-3 b: cdefg
2-9 c: ccccccccc`

Each line gives the password policy and then the password. The password policy indicates the lowest and highest number of times a given letter must appear for the password to be valid. For example, 1-3 a means that the password must contain a at least 1 time and at most 3 times.

In the above example, 2 passwords are valid. The middle password, cdefg, is not; it contains no instances of b, but needs at least 1. The first and third passwords are valid: they contain one a or nine c, both within the limits of their respective policies.

How many passwords are valid according to their policies?

In [39]:
# Prepare an empty list for our data
data = []

# Open the file as f, pull one line at a time, strip it of the \n at the end of each line, and append it to our data. 
with open("AoC2020_2.txt") as f:
    for line in f.readlines():
        data.append(line.strip())

'9-10 m: mmmmnxmmmwm'

### str.split approach

Split the string on the colon, separating the rule and the password, and only focus on the rule for now

In [13]:
rule = data[0].split(':')[0]

The last character of the rule is always the character of interest

In [20]:
key = rule[-1]

We can then further split on the space to get the number range, and then on the - to get the low and high limits

In [19]:
(low, high) = rule.split(' ')[0].split('-')

We now have a key, and low and high limits. We now just want to search the password and make sure the character features a correct amount of times. 

In [28]:
correct = 0
password = data[0].split(': ')[1]
count = password.count(key)
if count >= int(low) and count <= int(high):
    correct += 1

In [29]:
correct

0

This seems to work, as indeed the first entry is not obeying its policy. Now to standardise this and loop through the dataset:

In [35]:
correct = 0

for line in data:
    rule = line.split(':')[0]
    key = rule[-1]
    (low, high) = rule.split(' ')[0].split('-')
    
    password = line.split(': ')[1]
    count = password.count(key)
    if count >= int(low) and count <= int(high):
        correct += 1

In [36]:
correct

469

There are 469 passwords that obey their policies

### RegEx approach

In [4]:
import re

In [10]:
re.search(r'\d+-\d+ \w', data[0])[0]

'9-10 m'

## Part 2

----------

--- Part Two ---

While it appears you validated the passwords correctly, they don't seem to be what the Official Toboggan Corporate Authentication System is expecting.

The shopkeeper suddenly realizes that he just accidentally explained the password policy rules from his old job at the sled rental place down the street! The Official Toboggan Corporate Policy actually works a little differently.

Each policy actually describes two positions in the password, where 1 means the first character, 2 means the second character, and so on. (Be careful; Toboggan Corporate Policies have no concept of "index zero"!) Exactly one of these positions must contain the given letter. Other occurrences of the letter are irrelevant for the purposes of policy enforcement.

Given the same example list from above:

`1-3 a: abcde is valid: position 1 contains a and position 3 does not.
1-3 b: cdefg is invalid: neither position 1 nor position 3 contains b.
2-9 c: ccccccccc is invalid: both position 2 and position 9 contain c.`

How many passwords are valid according to the new interpretation of the policies?


-----------

For this we mostly reuse our code, but rather than counting the occurences of the key, we check the low and high position in the string and check if one of those is the key. 

In [32]:
# Prepare an empty list for our data
data = []

# Open the file as f, pull one line at a time, strip it of the \n at the end of each line, and append it to our data. 
with open("AoC2020_2.txt") as f:
    for line in f.readlines():
        data.append(line.strip())

In [33]:
correct = 0

for line in data:
    rule = line.split(':')[0]
    key = rule[-1]
    (low, high) = rule.split(' ')[0].split('-')
    
    password = line.split(': ')[1]
    if (password[int(low)-1] == key) ^ (password[int(high)-1] == key) :
        correct += 1

In [34]:
correct

267